### PR TITLE
Put api version in the high level client; add Group(), Version(), and GroupVersion() to Request

### PR DESCRIPTION
**Pull request by janetkuo**
Partially fixes - #14133

cc @lavalamp @caesarxuchao

### CODE REVIEW

**Comment by smarterclayton** [comment]
This is an incredible amount of redundant code - is this really the best way to do this?  Why not wrap the rest client interface with a group applier?

File: pkg/client/unversioned/request.go [hunk 1]
```diff
@@ -98,6 +98,13 @@ type Request struct {
 
 	apiVersion string
 
+	// group/version
+	group      string
+	version    string
```
Comment (deads2k) [inline pkg/client/unversioned/request.go]: How is this field different from `apiVersion` above?  Does this change the way that `baseURL` is used?  

Also, I remember an `UnversionedPath` method that did some crazy things.  This probably has effects tehre too.

**Comment by smarterclayton** [comment]
To offer a concrete suggestion instead of just complaining.  Instead of exposing this like that, have the embedded `RESTClient` in each client be an interface (we made it one in resource.Builder, and it's Get / Put / Post / Delete that returns a concrete request object).  When we init the client to an API version, wrap RESTClient Get / Put / Post / Delete with helpers that call the wrapped rest client and then set GroupVersion.  Then no client code needs to know about api version and we go back to a much cleaner code base:

```
type VersionedRESTClient struct {
  wrapped RESTClient
  groupVersion string
}

func (c *VersionedRESTClient) Get() *Request {
  return c.Get().GroupVersion(c.groupVersion)
}
```

**Comment by lavalamp** [comment]
@smarterclayton I want to supply typed versioned clients; that means the code has to be redundant anyway. I hope to auto-generate it eventually.

File: pkg/client/unversioned/request.go (hunk 1 above)
Comment (JanetKuo) [inline pkg/client/unversioned/request.go]: `apiVersion` is currently not used in Request. Both `baseURL` and `apiVersion` are derived from RESTClient when it creates a new Request, and neither `apiVersion` nor `group`/`version` affects `baseURL` in Request. This is the first step before we remove references to "version" from the RESTClient. 

`UnversionedPath` strips out the version word (such as `v1`) in `baseURL` and save that to Request `path` (doesn't change `baseURL`), and `path` and `baseURL` are used for generating a final url in URL().

File: pkg/client/unversioned/request_test.go [hunk 2]
```diff
@@ -729,6 +729,7 @@ func TestDoRequestNewWay(t *testing.T) {
 	defer testServer.Close()
 	c := NewOrDie(&Config{Host: testServer.URL, Version: testapi.Default.Version(), Username: "user", Password: "pass"})
 	obj, err := c.Verb("POST").
+		GroupVersion(testapi.Default.Version()).
```
Comment (caesarxuchao) [inline pkg/client/unversioned/request_test.go]: Why not use testapi.Default.GroupAndVersion()? Although they both return "v1", using GroupAndVersion() looks more consistent.

File: pkg/client/unversioned/request.go [hunk 3]
```diff
@@ -238,6 +246,35 @@ func (r *Request) AbsPath(segments ...string) *Request {
 	} else {
 		r.path = path.Join(segments...)
 	}
+	r.setAbspath = true
+	return r
+}
+
+// Group
+func (r *Request) Group(group string) *Request {
```
Comment (caesarxuchao) [inline pkg/client/unversioned/request.go]: Could you provide a use case of Group? When would someone set the group but not the version at the same time?

Comment (lavalamp) [inline pkg/client/unversioned/request.go]: If people have group & version as separate strings, we don't want to make
them concatenate.

File: pkg/client/unversioned/request_test.go (hunk 2 above)
Comment (JanetKuo) [inline pkg/client/unversioned/request_test.go]: Fixed.

**Comment by smarterclayton** [comment]
How so?  Initialization can wrap, at minimum.  My concern is groupVersion
is a property of a rest client for a single groupVersion, and the semantics
of each client vary based on group version, not *with* group version (for
things like naming, path structure, variable names).

This is making rest client more generic than it was intended to be.

On Sep 18, 2015, at 12:23 PM, Daniel Smith <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> I want to supply typed
versioned clients; that means the code has to be redundant anyway. I hope
to auto-generate it eventually.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/14149#issuecomment-141498089>
.

**Comment by smarterclayton** [comment]
If I move a group version to a new server process because I need two
separate binaries handling them, that should be a detail of the rest client
initialized into the client object.  If the client needs to know its target
version that's fine (read it from the request) - I just don't think that's
a detail that client fills in (like server URL, base prefix, auth tokens,
user agent headers, etc).

On Sep 18, 2015, at 12:23 PM, Daniel Smith <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> I want to supply typed
versioned clients; that means the code has to be redundant anyway. I hope
to auto-generate it eventually.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/14149#issuecomment-141498089>
.

**Comment by JanetKuo** [comment]
@k8s-bot test this again

**Comment by JanetKuo** [comment]
@k8s-bot test this please

**Comment by lavalamp** [comment]
> If I move a group version to a new server process because I need two separate binaries handling them, that should be a detail of the rest client initialized into the client object.

Do we really want to pass around a separate client for each group? I don't think that's very convenient.

I think we may have different models of how this should work. I was thinking every call should hit (an) apiserver, and then be redirected (proxied if necessary) to the final 3rd party binary. This lets us handle auth, admission, etc without trusting the 3rd party to implement this stuff correctly.

But even if we want the client to talk directly to the 3rd party binary, I still think that the RESTClient layer should handle that, or that we should pass around a meta-rest-client that selects a RESTClient based on the group.

> the semantics of each client vary based on group version

I guess don't see the point of this. If you don't want to follow the k8s api semantics, why do you want to register with apiserver?

File: pkg/client/unversioned/client.go [hunk 4]
```diff
@@ -199,3 +203,7 @@ func IsTimeout(err error) bool {
 func (c *Client) Experimental() ExperimentalInterface {
 	return c.ExperimentalClient
 }
+
+func ClientGroupVersion() string {
```
Comment (lavalamp) [inline pkg/client/unversioned/client.go]: I don't think this needs to be a function. Just make the constants public.

File: pkg/client/unversioned/client.go [hunk 5]
```diff
@@ -28,6 +28,10 @@ import (
 	"k8s.io/kubernetes/pkg/version"
 )
 
+const (
+	clientGroupVersion = "v1"
```
Comment (lavalamp) [inline pkg/client/unversioned/client.go]: These are constants that logically belong to the high level client. I'd prefer to see a list here, e.g.:

```go
const (
  CoreGroupVersion = "v1"
  ExperimentalGroupVersion = "experimental/v1alpha1"
  ...
)
```

For extra points, the typesafe version would look something like:

```go

type GroupVersion string

const (
  Core = GroupVersion("v1")
  Experimental = GroupVersion("experimental/v1alpha1")
  ...
)
```

This is nice, but means you'd want to set the type of the parameter the Request.GroupVersion() function takes.

File: pkg/client/unversioned/request.go (hunk 1 above)
Comment (lavalamp) [inline pkg/client/unversioned/request.go]: Please remove apiVersion if it's not used, that will reduce this confusion.

File: pkg/client/unversioned/request.go [hunk 6]
```diff
@@ -98,6 +98,13 @@ type Request struct {
 
 	apiVersion string
 
+	// group/version
```
Comment (lavalamp) [inline pkg/client/unversioned/request.go]: Add that these are set by the Group/Version/GroupVersion methods.

Comment (lavalamp) [inline pkg/client/unversioned/request.go]: (So people don't expect them to have been set by the constructor.)

File: pkg/client/unversioned/request.go [hunk 7]
```diff
@@ -238,6 +246,35 @@ func (r *Request) AbsPath(segments ...string) *Request {
...
+func (r *Request) Version(version string) *Request {
+	r.version = version
+	r.setVersion = true
+	return r
+}
+
+// GroupVersion reads in a group/version string and set the group and version
+func (r *Request) GroupVersion(groupVersion string) *Request {
+	groupAndVersion := strings.Split(groupVersion, "/")
+	if len(groupAndVersion) == 1 {
```
Comment (lavalamp) [inline pkg/client/unversioned/request.go]: I think @caesarxuchao has a utility function somewhere that does this-- should just call it if possible.

Comment (caesarxuchao) [inline pkg/client/unversioned/request.go]: https://github.com/kubernetes/kubernetes/blob/master/pkg/api/util/group_version.go

File: pkg/apis/experimental/v1/types_swagger_doc_generated.go [hunk 8]
```diff
@@ -259,8 +259,8 @@ func (ResourceConsumption) SwaggerDoc() map[string]string {
...
+	"maxSurge":        "The maximum number of pods that can be scheduled above the desired number of pods. Value can be an absolute number (ex: 5) or a percentage of desired pods (ex: 10%). This can not be 0 if MaxUnavailable is 0. Absolute number is calculated from percentage by rounding up. By default, a value of 1 is used. Example: when this is set to 30%, the new RC can be scaled up immediately when the rolling update starts, such that the total number of old and new pods do not exceed 130% of desired pods. Once old pods have been killed, new RC can be scaled up further, ensuring that total number of pods running at any time during the update is atmost 130% of desired pods.",
```
Comment (bgrant0607) [inline pkg/apis/experimental/v1/types_swagger_doc_generated.go]: Why is this update in this PR?

cc @nikhiljindal

Comment (caesarxuchao) [inline pkg/apis/experimental/v1/types_swagger_doc_generated.go]: HEAD is broken. I sent #14213 to fix.

**Comment by bgrant0607** [comment]
From this comment:
https://github.com/kubernetes/kubernetes/pull/14149#issuecomment-141498089

I infer that this will all change again as part of the move to versioned clients?

**Comment by bgrant0607** [comment]
It doesn't have to block this PR, but sometime soon we should write up a proposal of what we want the client layering to be.

**Comment by smarterclayton** [comment]
On Sep 18, 2015, at 7:08 PM, Daniel Smith <notifications@github.com> wrote:

If I move a group version to a new server process because I need two
separate binaries handling them, that should be a detail of the rest client
initialized into the client object.

Do we really want to pass around a separate client for each group? I don't
think that's very convenient.

I think we may have different models of how this should work. I was
thinking every call should hit (an) apiserver, and then be redirected
(proxied if necessary) to the final 3rd party binary. This lets us handle
auth, admission, etc without trusting the 3rd party to implement this stuff
correctly.



The problem is in practice that bottlenecks every single api though the
apiservers, and in some cases you will have clients in those other Apis
calling back to that proxy and back through.  Also, if you terminate authn
and authz at the first apiserver, how is the second apiserver secured?  How
does it take policy decisions that the first apiserver is not aware of?
There are concrete uses for that today that are insufficiently covered by
that proposal.  Brendan's third party resource is more in keeping with that
proxy model, a real  API like Openshift today would have significant
challenges.

We've talked about this a lot but we haven't *designed* it - my working
assumption right now is that we can't do this properly until we design a
solution.  The assumption the RESTClient object is designed around today is
that it encapsulated all details to speak to a specific API version - all
other details above that are details of a higher level abstraction (RestMapping
and ClientForMapping in the CLI Factory)

But even if we want the client to talk directly to the 3rd party binary, I
still think that the RESTClient layer should handle that, or that we should
pass around a meta-rest-client that selects a RESTClient based on the group.


RESTClient today selects the apiversion.  I don't want to change that one
day before feature freeze because that has massive tricke down impacts into
other code domains.  If RESTClient abstracts group and version there's no
impact (as I suggested).


the semantics of each client vary based on group version

I guess don't see the point of this. If you don't want to follow the k8s
api semantics, why do you want to register with apiserver?



Client library - for instance, an object rename (minions to nodes).  Back
compat method checking (trying post before get on exec).  What arguments
are supported.  RESTClient is our generic "supports API semantics" pattern
(and even then we could vary URL construction by having different
RESTClient impls that are read by looking at URL patterns returned by
swagger, for instance).  Each client interface maps that rest client to a
specific set of inputs to a restful model (reosurce, name, fields).

This, incidentally, is why field conversion does not belong in request.go
but in each ClientInterface, because request.go is a particular target for
a "generic" API pattern and fields are tied to a kind.  But that's not
critical right now.


—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/14149#issuecomment-141590450>
.

**Comment by smarterclayton** [comment]
Comment updated to fix horrible gmail->github format bomb

**Comment by smarterclayton** [comment]
We should sync up at first opportunity.

**Comment by JanetKuo** [comment]
@k8s-bot test this again

**Comment by lavalamp** [comment]
@smarterclayton OK I don't think our positions are actually that far apart.

My mental model of the client has (at least) these layers:

Lowest: config/certs/server location
RESTClient-- building URLs with helper functions, encoding/decoding api objects, making actual HTTP calls
Typed go client: the `c.Core().Pods("foo").List() (*v1.PodList, error)` style functions.

I guess you want to insert a MultiServerRESTClient; to use this, the high level client would say something like `c.Destination("group/version").Get()...` instead of my `c.Get().GroupVersion("group/version")...`.

My requirements are that 1. we pass around a single object everywhere (I do NOT want to see any more `type x struct {Client *client, ExperimentalClient *client}` structures) and 2. the high-level typed client needs to be able to specify the group/version, because the typed client is where the URLs are constructed to match the go types.

kubectl acts like a high level client. It needs to do version negotiation and then generically support some operations against every group/version/resource in addition to more complex operations against types it directly understands.

I don't particularly care if we add a MetaRESTClient, as opposed to making the existing rest client accept multiple configs/certs/destinations.

**Comment by smarterclayton** [comment]
On Sep 21, 2015, at 12:43 PM, Daniel Smith <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> OK I don't think our
positions are actually that far apart.

My mental model of the client has (at least) these layers:

Lowest: config/certs/server location
RESTClient-- building URLs with helper functions, encoding/decoding api
objects, making actual HTTP calls
Typed go client: the c.Core().Pods("foo").List() (*v1.PodList, error) style
functions.

I guess you want to insert a MultiServerRESTClient; to use this, the high
level client would say something like
c.Destination("group/version").Get()... instead of my
c.Get().GroupVersion("group/version")....



Today, I would expect someone to construct a RESTClient that points to a
group version, then wrap the Client or ExperimentalClient around the
RESTClient.  I would expect client.New(config) to return an
ExperimentalClient and Client as our static path.  I would expect to add a
new client.Discover(groups) which used /apis/ to return a map of group ->
RESTClient on preferred versions.  I might expect
client.NewDiscover(config) that used Discover, then New.

I do want to preserve all the static paths for now because we use them
heavily on OS.


My requirements are that 1. we pass around a single object everywhere (I do
NOT want to see any more type x struct {Client *client, ExperimentalClient
*client} structures) and 2. the high-level typed client needs to be able to
specify the group/version, because the typed client is where the URLs are
constructed to match the go types.



I think of Request as abstracting URL construction (it does today).  I
expect that Client, if they need to, can get at groupVersion in order to
pass slightly different parameters (resource name).  I think Request could
have helpers injected into it that do type mapping, but I think RESTClient
shouldn't have any direct references to scheme or the Kube objects.

I agree the struct is awkward, Openshift just passed around two clients.
Callers are expected to know whether they need a Kube client or OS client.


kubectl acts like a high level client. It needs to do version negotiation
and then generically support some operations against every
group/version/resource in addition to more complex operations against types
it directly understands.



I think eventually kubectl should call client.Discover and get the map of
groups -> rest clients, and also get the list of resources that map to
those groups, and it uses RESTClient directly (like it has today) with
resource.Helper.  RESTClient has to abstract enough URL construction that
resource.Helper just works.


I don't particularly care if we add a MetaRESTClient, as opposed to making
the existing rest client accept multiple configs/certs/destinations.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/14149#issuecomment-142038337>
.

**Comment by smarterclayton** [comment]
One note, RESTClient *must* correctly build URLs that are based on group
and version, so GroupVersion *should* be hidden behind RESTClient

**Comment by lavalamp** [comment]
Sorry about the wild goose chase, Janet :)

**Pull request closed without merging**